### ISSUE TITLE
Building R CRAN packages with Fortran code using "fpm" is not possible, but should be!

### ISSUE THREAD

**Issue by AndreMikulec** [fortran-lang/fpm#336]
All,

First,

Thank you for the wonderful work on the "fpm".

Second,

In the manual "Writing R Extensions"
In the section "1.2.3 Using F9x code"
https://cran.r-project.org/doc/manuals/r-release/R-exts.html
reads
```
Modern versions of Fortran support modules, 
whereby compiling one source file creates a module file which is then included in others. 
(Module files typically have a .mod extension: they do 
depend on the compiler used and so should never be included in a package.) 
This creates a dependence which make will not know about 
and often causes installation with a parallel make to fail
```

But the page https://github.com/Fortran-lang/fpm/blob/master/PACKAGING.md
reads in the section "Single-module library"
```
fpm first ran gFortran to emit the binary object (math_constants.o) 
and module (math_constants.mod) files. 
. . .
Uniquely and exactly map a source file (.f90) to its object (.o) and module (.mod) files.
```

In conclusion, 

May one consider creating the modules with the .f90 extension (instead of the .mod extension), 
so in the future, one can use "fpm" in the processes of building 
R CRAN package that contains Fortran source code.

Fortran is found in many places in the 20,000 R CRAN packages.
This could be a great area of expansion for the "fpm".

If "fpm" can get compatible as a helper method to build R CRAN packages that contain Fortran code,
then this situation may be a big boost to  "fpm" (and Fortran).

Here are the Fortran files found in R CRAN packages.

.f -> 4,192 hits
https://github.com/search?q=org%3Acran+extension%3Af
.f90 -> 881 hits
https://github.com/search?q=org%3Acran+extension%3Af90
.f95 -> 156 hits
https://github.com/search?q=org%3Acran+extension%3Af95

Andre Mikulec
Andre_Mikulec@Hotmail.com

**Comment by certik** [comment]
Hi @AndreMikulec, we definitely want `fpm` to work with the R ecosystem. Can you expand on your proposal? I don't understand how it would help to rename `.mod` to `.f90`, or why that is a good idea.

In general, our approach is to first figure out a good workflow that makes sense for Fortran, and then work with other communities such as Spack, Conda, R, etc., how to make it work.

**Comment by LKedward** [comment]
Hi @AndreMikulec, thanks for opening this issue.
Regarding Fortran modules, *fpm* automatically works out the inter-module dependencies and builds objects in the correct order and in parallel where possible, so there is no need to worry about the `.mod` files.

I am not familiar with R CRAN packages and how they are built, however I see no fundamental reason why *fpm* can't already be used to build the Fortran dependencies. It looks like the CRAN packages are in a compatible folder structure for *fpm*. 

After a brief look it looks like CRAN packages cannot have multiple Fortran modules in different files since, as you mention, module dependencies are not accounted for currently; using *fpm* would solve this issue.

**Comment by AndreMikulec** [comment]
### First

".mod" is not-a-file-extension, of a file, that contains Fortran code.

File extensions
http://fortranwiki.org/fortran/show/File+extensions

The ".mod" file extension is a "compiler specific thing".  So, this not a  "Fortran Package Manager?  So, is this a "Compiler-Specific-Fortran Package Manager"? 

A quick discussion is here:
```
A primary problem for distributing a compiled library for a specific platform
is that module files are compiler-specific.
```
http://fortranwiki.org/fortran/show/Library+distribution

### Second

In R,  to create a "dll" that is loadable into R, one does the following:

R CMD SHLIB [options] files | linker options

Therefore, one does, the following to create a .dll on windows (or to create an equivalent shared object on Linux, Mac, or Solaris).

R CMD SHLIB <options> *.f90 <link options>

(The command to create (and install) an R CRAN package that contains compiled Fortran is R CMD INSTALL (and this is just a thick wrapper (and more)  over 'R CMD SHLIB . . .'.

In the Fortran packaging system (if, it is, to be able to work with R Fortran compiling):

"R" would replace "gfortran".  "CMD SHLIB options" would be, of course, options.  "link options" would be options, of course, sent to the linker.  The rest is controlled by either variables that were set when R was compiled or variables set in the environment where "R CMD SHLIB . . ." is ran.

### Summary

What I am, essentially saying, is that the packaging system should more "flexible" than what is there now.

**Comment by awvwgk** [comment]
@AndreMikulec Thank you for your input on this.

From my experience of interfacing Fortran projects with C and Python, the process of creating an extension for another language is quite orthogonal to reusing Fortran packages/modules in Fortran projects. Currently fpm focus primarily on the latter, but we want to make fpm more flexible to allow interfacing with other languages as well.

If I read the R CRAN documentation on extensions correctly a Fortran extension module must be bound either via `iso_c_binding` as C external to R or as non-module procedure using a Fortran 77 style implicit interface. This is quite similar to the approach used to bind to Python. In any case this requires to export a shared object, but in general no special attention beside include paths or link line arguments. Building shared objects is something fpm currently can't do, but should be able to do at some point.

From the R CRAN documentation I get the impression that extensions all builds around the autoconf intrastructure and rely pretty heavily on automake and m4 features, but correct me if I'm wrong. Comparing this again to Python which relies heavily on setuptools to build extensions, I found that building Python extensions without setuptools becomes much easier.

In summary, we first have to give fpm some basic intercompatibility functionality and than build on-top of this more elaborated language specific extensions.

**Comment by AndreMikulec** [comment]
R CMD SHLIB does not provide any infrastructure to interface Fortran to another language.  Interfacing to another language is not the objective.  Interfacing to another language (if the R package author wishes to do so), would be a function of the chosen compiler (of which can actually be many different compilers and OS platforms) and/or handwritten code (or code generated from an external helper tool) by the author.

The "autoconf intrastructure" does not matter.  The point is that, the situation would be useful to be able to replace out "gfortran" or "ifort" or "whatever", using "R CMD SHLIB", and replace out ".mod" files with just ".f90" files.  

The replacement of gfortran" or "ifort" or "whatever", using "R CMD SHLIB" should not be (obviously) too hard.

However, leaving "fpm" hardcoded to use only ".mod" files (to store Fortran module code) would eliminate the ability of a user of "fpm" to use R as an "alternate"(specific) compiler.

**Comment by awvwgk** [comment]
I think we are talking about different things here. There might be a confusion between shared modules and Fortran module files here. The creation of the shared module, which can be eventually loaded into R, does not depend on the Fortran module files.

Creating a shared module is independent of the actual compilation of the Fortran source code to object files. Replacing the compiler with `R CMD SHLIB` would not yield the expected result in the context of fpm.

The important step for creation of the shared module from the already available object file following the naming convention required by R from the fpm side.
